File: it/lift_basic/src/main/java/OutsideInService.java


import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.security.NoSuchAlgorithmException;
import java.util.List;

import org.apache.commons.codec.binary.Hex;
import org.apache.commons.codec.digest.DigestUtils;

import com.google.gson.Gson;


public class OutsideInService {

	public static final String API_KEY = "66gs4kyjqcgptdf78d4ax836";
	public static final String SHARED_SECRET = "HfH7DEwCdt";
	
	public static Stories getStories(String uuid) throws IllegalStateException, UnsupportedEncodingException, NoSuchAlgorithmException, IOException {
		String json = HTTPHelper.getHTML(getBaseURL() + uuid + "/stories?" + getSignature()).toString();
		Gson gson = new Gson();
		return gson.fromJson(json, Stories.class);
	}
		
	public static Location getLocation(String query) throws IllegalStateException, UnsupportedEncodingException, NoSuchAlgorithmException, IOException {
		String json = HTTPHelper.getHTML(getBaseURL() + "named/" + query + "?" + getSignature()).toString();
		Gson gson = new Gson();
		Locations locations = gson.fromJson(json, Locations.class);
		if(locations.getLocations() != null && locations.getLocations().size() > 0)
			return locations.getLocations().get(0);
		return null;
	}
	
	public static String getBaseURL() {
		return "http://hyperlocal-api.outside.in/v1.1/locations/";
	}
	
	public static String getSignature() throws UnsupportedEncodingException, NoSuchAlgorithmException {
		long unixTime = System.currentTimeMillis() / 1000L;
		new Hex();
		String md5 = Hex.encodeHexString(DigestUtils.md5((API_KEY + SHARED_SECRET + String.valueOf(unixTime))));
		return "dev_key=" + API_KEY + "&sig=" + md5;
	}
	
	public static void main(String[] args) throws NoSuchAlgorithmException, IllegalStateException, IOException {
		System.out.println(getStories(getLocation("california").uuid));
	}
	
	class Locations {
		private List<Location> locations;

		public Locations(List<Location> locations) {
			this.locations = locations;
		}

		public List<Location> getLocations() {
			return locations;
		}

		public void setLocations(List<Location> locations) {
			this.locations = locations;
		}
		
		public String toString() {
			return locations.toString();
		}
		
	}
	
	class Location {
		private String uuid;
		private String lat;
		private String lng;

		public Location(String uuid, String lat, String lng) {
			this.uuid = uuid;
			this.lat = lat;
			this.lng = lng;
		}
		
		public String getLat() {
			return lat;
		}

		public void setLat(String lat) {
			this.lat = lat;
		}

		public String getLng() {
			return lng;
		}

		public void setLng(String lng) {
			this.lng = lng;
		}
		
		public String getUuid() {
			return uuid;
		}

		public void setUuid(String uuid) {
			this.uuid = uuid;
		}
		
		public String toString() {
			return uuid;
		}
	}
	
	class Stories {
		private List<Story> stories;

		public Stories(List<Story> stories) {
			this.stories = stories;
		}

		public List<Story> getStories() {
			return stories;
		}

		public void setStories(List<Story> stories) {
			this.stories = stories;
		}
		
		public String toString() {
			String result = "";
			for(Story story : stories) {
				result += story.toString() + "\n";
			}
			return result;
		}
	}
	
	class Story {
		private String story_url;

		private String title;
		private String uuid;
		private String summary;
		
		public Story(String story_url, String title, String uuid,
				String summary, String published_at) {
			this.story_url = story_url;
			this.title = title;
			this.uuid = uuid;
			this.summary = summary;
			this.published_at = published_at;
		}
		
		private String published_at;
		
		public String getStory_url() {
			return story_url;
		}
		public void setStory_url(String story_url) {
			this.story_url = story_url;
		}
		public String getTitle() {
			return title;
		}
		public void setTitle(String title) {
			this.title = title;
		}
		public String getUuid() {
			return uuid;
		}
		public void setUuid(String uuid) {
			this.uuid = uuid;
		}
		public String getSummary() {
			return summary;
		}
		public void setSummary(String summary) {
			this.summary = summary;
		}
		public String getPublished_at() {
			return published_at;
		}
		public void setPublished_at(String published_at) {
			this.published_at = published_at;
		}
		
		@Override
		public String toString() {
			return "Story: \n\tstory_url: " + story_url + "\n\ttitle: " + title
					+ "\n\tuuid: " + uuid + "\n\tsummary: " + summary
					+ "\n\tpublished_at: " + published_at + "";
		}
	}
}

File: j/app/OutsideInService.scala
package apis

import dispatch._

import org.apache.commons.codec.binary.Hex;
import org.apache.commons.codec.digest.DigestUtils;

/*
// Parse JSON arrays
parse[List[Int]]("[1,2,3]") //=> List(1,2,3)

// Parse JSON objects
parse[Map[String, Int]]("""{"one":1,"two":2}""") //=> Map("one"->1,"two"->2)

// Parse JSON objects as case classes
// (Parsing case classes isn't supported in the REPL.)
case class Person(id: Long, name: String)
parse[Person]("""{"id":1,"name":"Coda"}""") //=> Person(1,"Coda")

// Parse streaming arrays of things
for (person <- stream[Person](inputStream)) {
  println("New person: " + person)
}

         "category": {
                "display_name": "Sub-Metro Area",
                "name": "sub_metro"
            },
            "city": "Training",
            "display_name": "Chicago",
            "hot_for_requested": false,
            "lat": 41.838272094726562,
            "lng": -87.690010070800781,
            "state": "Illinois",
            "state_abbrev": "IL",
            "url": "http://outside.in/chicago-chicagoland-il?utm_content=Chicago&utm_source=API_OutsideIn_0&utm_medium=API&utm_campaign=APILinkback",
            "url_name": "chicago-chicagoland-il",
            "uuid": "e9cf5975-9c01-4fbb-bebd-ab004ff20616"
        },

*/

case class Location(
  display_name: String,
  state: String,
  url: String,
  lng: String,
  lat: String,
  uuid: String,
  city: String,
  category: Map[String, String],
  hot_for_requested: String,
  state_abbrev: String,
  url_name: String
)

case class Locations(total: String, locs: List[Location])

object OutsideInService {

  val API_KEY = "66gs4kyjqcgptdf78d4ax836"
  val SHARED_SECRET = "HfH7DEwCdt"

  def getStories(query: String): Iterable[(String, String)] = {
    val uuid = getLocation(query)
    val json = Http(url(getBaseURL() + uuid + "/stories?" + getSignature()) as_str)
    println(json)
    List()
  }

  def getLocation(query: String): String = {
    val json = Http(url(getBaseURL() + "named/" + query + "?" + getSignature()) as_str)
    val JO = """"uuid":"(.*)".*""".r
    val o = json.split(',')
    val x = o foreach { case JO(uuid) => return uuid }
    ""
  }

  def getBaseURL() = "http://hyperlocal-api.outside.in/v1.1/locations/"

 def getSignature() = {
    val unixTime = System.currentTimeMillis() / 1000L
    new Hex()
    val md5 = Hex.encodeHexString(DigestUtils.md5((API_KEY + SHARED_SECRET + String.valueOf(unixTime))))
    "dev_key=" + API_KEY + "&sig=" + md5
  }

}
